### DISCUSSION TITLE
Histogram of LCH image

### DISCUSSION THREAD

**Discussion by schmijos** [libvips/libvips#2520] [Q&A] [answered]
I'm trying to get an ndim histogram of the CH part of an image:

(here's the image: ![sample](https://user-images.githubusercontent.com/245443/140624693-26474849-c133-4b24-a8b7-0d3e4b716993.png))

```rb
#!/usr/bin/ruby
require "vips"

img = Vips::Image.new_from_file('spec/fixtures/files/sample.png', access: :sequential).colourspace(:lch)

img[1..2].to_a
=> [[[102.3130874633789, 39.18145751953125],
  [101.67424774169922, 39.16696548461914],
  [0.4031951427459717, 110.17615509033203],
  [118.2603988647461, 135.60243225097656]],
 [[102.3130874633789, 39.18145751953125],
  [102.67890930175781, 39.159664154052734],
  [118.54254913330078, 135.6748809814453],
  [0.00013431724801193923, 326.3099365234375]],
 [[133.49240112304688, 306.25054931640625],
  [0.33749398589134216, 201.97052001953125],
  [115.21880340576172, 328.395263671875],
  [95.71614074707031, 102.8958511352539]],
 [[0.019027704373002052, 326.1108093261719],
  [132.93775939941406, 306.2183532714844],
  [95.6482925415039, 102.39655303955078],
  [96.01411437988281, 102.91535186767578]]]

img[1..2].hist_find_ndim(bins: 5).to_a
=> [[[16], [0], [0], [0], [0]],
 [[0], [0], [0], [0], [0]],
 [[0], [0], [0], [0], [0]],
 [[0], [0], [0], [0], [0]],
 [[0], [0], [0], [0], [0]]]
```

Why are all values counted into 0,0,0?

The documentation says something about casting ([`hist_find_ndim`](https://www.libvips.org/API/current/libvips-arithmetic.html#vips-hist-find-ndim)), but still most of the values should be in range, shouldn't they?

**Comment by jcupitt** [comment, marked as answer]
Hi @schmijos,

You are taking the histogram of a float image, so it's casting it to uint16 before scanning. Five bins on a range of 0 - 65535 means each bin is more than 10,000 wide, so all your values fall in bin 0. If you cast to uint8 (`uchar`) yourself, you should see something more expected.

It's best to fix the range, of course. C is roughly -128 to 128, and h is 0 - 360, so you can move them to 0 - 255 like this:

```ruby
#!/usr/bin/ruby

require "vips"

x = Vips::Image.new_from_file ARGV[0]
x = x.colourspace("lch")[1..2]
x = (x * [1, 256 / 360] + [128, 0]).cast("uchar")
arr = x.hist_find_ndim(bins: 5).to_a
puts "histogram: #{arr}"
```

I see:

```
$ ./schmijos.rb ~/pics/tiny.png
histogram: [[[0], [0], [4], [0], [12]], [[0], [0], [0], [0], [0]], [[0], [0], [0], [0], [0]], [[0], [0], [0], [0], [0]], [[0], [0], [0], [0], [0]]]
```

**Reply by schmijos** [reply]
Thank you very much! This makes sense.

Nevertheless I find it a bit disturbing that there's no way around touching every value before counting them up (a read-only operation).

**Reply by jcupitt** [reply]
I'm not sure I understand. Could you explain what you mean by "touching every value"?  The `x = x.yyy` isn't modifying anything, I'm just reusing the variable name.

**Reply by schmijos** [reply]
Hm… then I maybe was a bit fast on judging `(x * [1, 256 / 360] + [128, 0]).cast("uchar")`?
Isn't this a matrix multiplication running through all the values? Aren't we doing double the work than would be needed to count up an RGB image?

**Reply by jcupitt** [reply]
It's just doing multiply-add (a linear transform) on each pixel value to change the range. The conversion to LCh is much more expensive! libvips is lazy and on demand, you you can maybe think of the linear transform as an extra stage added to the front of the histogram accumulator.

You could convert to HSV instead of LCh. It has 0-255 for each channel so there's no rescaling necessary. Something like:

```
x = Vips::Image.new_from_file ARGV[0]
x = x.colourspace("hsv")[1..2]
arr = x.hist_find_ndim(bins: 5).to_a
puts "histogram: #{arr}"
```

It'd be a lot quicker, since you skip the LCh conversion, but of course also much less accurate.

**Reply by jcupitt** [reply]
Or another optimisation might be to call `linear` directly. This computes `y = x * a + b`, and has an option for `uchar` output (so you can skip the cast). Perhaps:

```ruby
x = Vips::Image.new_from_file ARGV[0]
x = x.colourspace("lch")[1..2]
x = x.linear [1, 256 / 360], [128, 0], uchar: true
arr = x.hist_find_ndim(bins: 5).to_a
puts "histogram: #{arr}"
```

I don't know if that would be measurably quicker.

**Reply by schmijos** [reply]
Thank you for those thoughtful answers! You're putting so much care into this library. Can I send you a beer somehow?

**Comment by jcupitt** [comment]
(perhaps the histogram functions should always cast to uchar, with ushort as an option you have to turn on explcitly? It might be less surprising ... though it's too late to change the behaviour now, unfortunately)

**Comment by jcupitt** [comment]
I've tried to clarify the doc comments.
